drivers/key/key_gpio.c

```diff
@@ -0,0 +1,343 @@
+ */
+
+#define DT_DRV_COMPAT gpio_keys
+
+/**
+ * @file
+ * @brief GPIO driven KEYs
+ */
+
+#include <string.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/key.h>
+#include <zephyr/dt-bindings/key/key.h>
+#include <zephyr/zephyr.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(key_gpio);
+
+#ifndef MAX
+#define MAX(x, y) (((x) > (y)) ? (x) : (y))
+#endif
+
+#define SCAN_INTERVAL CONFIG_KEY_GPIO_SCAN_INTERVAL
+#define TIME_DEBOUNCE CONFIG_KEY_GPIO_TIME_DEBOUNCE
+#define TIME_LONG     CONFIG_KEY_GPIO_TIME_LONG
+#define TIME_HOLD     CONFIG_KEY_GPIO_TIME_HOLD
+
+/**
+ * @brief KEY information structure
+ *
+ * This structure gathers useful information about KEY controller.
+ *
+ * @param label KEY label.
+ * @param code  KEY code.
+ * @param ms_debounce KEY debounce time
+ * @param ms_long KEY long time
+ * @param ms_hold KEY hold time
+ */
+struct key_info_dt_spec {
+	const char *label;
+	uint16_t code;
+	uint16_t ms_debounce;
+	uint16_t ms_long;
+	uint16_t ms_hold;
+};
+
+/**
+ * @brief Get KEY INFO DT SPEC
+ *
+ * @param node_id devicetree node identifier
+ */
+#define KEY_INFO_DT_SPEC_GET(node_id)                                 \
+{                                                                     \
+	.label = DT_LABEL(node_id), .code = DT_PROP_OR(node_id, code, 0), \
+	.ms_debounce = DT_PROP_OR(node_id, ms_debounce, 0),               \
+	.ms_long = DT_PROP_OR(node_id, ms_long, 0),                       \
+	.ms_hold = DT_PROP_OR(node_id, ms_hold, 0),                       \
+}
+
+typedef enum {
+	KEY_STATE_NONE,
+	KEY_STATE_PRESSED,
+	KEY_STATE_LONG_PRESSED,
+	KEY_STAT_MAXNBR
+} key_state_t;
+
+struct key_gpio_driver {
+	struct gpio_callback gpio_data;
+
+	void *pdata;
+	bool pressed;
+	key_state_t state;
+	key_event_t event;
+
+	uint32_t count_debounce;
+	uint32_t count_cycle;
+};
+
+struct key_gpio_data {
+	const struct device *dev;
+
+	key_callback_t callback;
+	struct k_work_delayable delayed_work;
+
+	struct key_gpio_driver *driver;
+};
+
+struct key_gpio_config {
+	uint8_t num_keys;
+	const struct gpio_dt_spec *gpio;
+	const struct key_info_dt_spec *info;
+};
+
+#define MS_TO_CYCLE(ms, default)      (((ms) ? (ms) : (default)) / SCAN_INTERVAL)
+#define KEY_EVENT_CODE(code, default) ((code != KEY_RESERVED) ? code : default)
+#define KEY_EVENT_CALL(dev, code, event)  \
+{                                         \
+	if (data->callback) {                 \
+		data->callback(dev, code, event); \
+	}                                     \
+}
+
+static bool key_gpio_one_key_proc(const struct device *dev, int idx)
+{
+	const struct key_gpio_config *config = dev->config;
+	struct key_gpio_data *data = dev->data;
+
+	const struct gpio_dt_spec *gpio = &config->gpio[idx];
+	const struct key_info_dt_spec *info = &config->info[idx];
+	struct key_gpio_driver *driver = &data->driver[idx];
+
+	bool pressed = (gpio_pin_get(gpio->port, gpio->pin) ^ gpio->dt_flags);
```

Comment:
This driver works great for me but this line causes pressed to always be true and so each event only ever fires once. This is on an stm32l4